src/workerd/io/actor-cache.c++

```diff
@@ -2843,28 +2853,25 @@ kj::Promise<void> ActorCache::flushImplUsingTxn(
     };
 
     for (auto& promise : promises) {
-      // Reuse `countDeleted` since it's already in a state object anyway.
-      rpcCountedDelete.countedDelete->countDeleted += co_await promise;
+      // TODO(now): Note to self - Are we absolutely, positively certain that we will always come
```

Comment:
I agree this looks likely to be buggy. Specifically, it looks most suspicous in the case where there are multiple `rpcDeletes` attached to the `rpcCountedDelete` and the first one succeeds but a latter one fails. There may be other scenarios where a retry could cause problems too, but that one jumps out.

It'd be interesting to try to write a test case to break this. I think it could be fixed by accumulating the counts from all the promises in a local variable first, and only adding them to `rpcCountedDelete.countedDelete->countDeleted` when they've all resolved successfuly. And then we could add them to the count at the same time that we set `rpcCountedDelete.countedDelete->completedInTransaction` (I don't actually see why we need to set that in callback below rather than immediately before returning from this lambda), which would make this look much less scary.

Reply:
https://github.com/cloudflare/workerd/pull/1915/commits/694b9515a38e8b8ac2a330c216362cc5aa76fa5d#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2841